docs/tutorialkit.dev/src/content/docs/guides/using-low-level-apis.mdx

```diff
@@ -0,0 +1,189 @@
+    const files = tutorialStore.documents.value;
+    const index = files["/index.js"].value as string;
+
+    if (index.length === 0) {
+      setMessage("Your index.js should not be empty");
+    } else if (!index.includes("export default")) {
+      setMessage("Your index.js should have a default export");
+    } else {
+      setMessage("All good");
+    }
+  }
+
+  return (
+    <>
+      <button
+        onClick={onClick}
+        className="px-4 py-1 rounded-md bg-tk-elements-primaryButton-backgroundColor text-tk-elements-primaryButton-textColor"
+      >
+        Help
+      </button>
+
+      {message && (
+        <Dialog title="Help" confirmText="OK" onClose={() => setMessage(null)}>
+          {message}
+        </Dialog>
+      )}
+    </>
+  );
+}
+```
+
+```mdx title="content.mdx"
+---
+type: lesson
+title: Lesson Topic
+---
+
+import { HelpButton } from "@/components/HelpButton";
+
+# Lesson Topic
+
+In this lesson we'll learn about...
+Click this button if you need help: <HelpButton client:load />
+```
+
+## Tutorial Core
+
+You can access Tutorial Core by importing the `tutorialkit:core` entrypoint.
+
+```ts
+import * as core from "tutorialkit:core";
+```
+
+The Tutorial Core provides access to webcontainer APIs. You can use it for example to read and modify files on the virtual file system.
+
+```ts
+import * as core from "tutorialkit:core";
+
+const webcontainer = await core.webcontainer;
+const contents = await webcontainer.fs.readdir("/", "utf8");
+
+console.log(contents);
+[
+  "index.html",
+  "node_modules",
+  "package-lock.json",
+  "package.json",
+  "vite.config.ts"
+]
+```
+
+### Example usage in React component
+
+Here is an example of custom React component that looks for updates in `/index.js` and provides user information about their lesson status.
+
+<video src={taskListVideo} controls />
+
+```tsx title="src/components/TaskList.tsx"
+import * as core from "tutorialkit:core";
+import { useEffect, useState } from "react";
+
+export function TaskList() {
+  const [hasDefaultExport, setDefaultExport] = useState<boolean>(false);
+  const [hasSumFunction, setSumFunction] = useState<boolean>(false);
+
+  useEffect(() => {
+    const unsubscribe = subscribeFileChanges("/index.js", (content) => {
+      // Naive text based checking just for example. Use proper AST analysis in real lessons.
+      setDefaultExport(content.includes("export default"));
+      setSumFunction(content.includes("function sum"));
+    });
+
+    return () => unsubscribe();
+  }, []);
+
+  return (
+    <ol>
+      <li>Add default export to index.js {hasDefaultExport ? "✅" : "❌"}</li>
+      <li>Add <code>function sum</code> to index.js {hasSumFunction ? "✅" : "❌"}</li>
+    </ol>
+  );
+}
+
+function subscribeFileChanges(filename: string, callback: (content: string) => void) {
+  const controller = new AbortController();
+
+  core.webcontainer.then((webcontainer) => {
+    const watcher = webcontainer.fs.watch(filename, async () => {
```

Comment:
We should recommend using `tutorialStore.onDocumentChanged` instead of recommending hooking into webcontainer. Doing this should always be last resort for users that know what they're doing.

First because it's going to be more efficient as it doesn't go through webcontainer.

But second, also because if the file is removed and then re-added, this would get stuck (but it should work with `onDocumentChanged` and if it doesn't then it's a bug).

Reply:
This was the only example I could think of from the `tutorialkit:core` module 😅 

But sure, happy to change this example with something else. As long as it's using `tutorialkit:core` which is the point here.

Reply:
Oh sorry I completely missed the goal of that example.

For an example using `tutorialkit:core`, how about we have one spawning a process that "checks" whether or not the lesson was solved (by running a test) and if the process exit with zero we show a confettis?